src/r-bridge/lang-4.x/ast/parser/json/format.ts

```diff
@@ -51,3 +76,47 @@ export function prepareParsedData(data: string): Entry[] {
 
 	return roots
 }
+
+export function convertPreparedParsedData(roots: CsvEntry[]): JsonEntry {
+	// Locate start, end of source file (order children in advance).
+	const rootEntries = roots.sort(orderOf)
+	const start = rootEntries[0]
+	const end = rootEntries[rootEntries.length - 1]
+
+	// Construct CsvEntry for the root, handling empty input.
+	const csvParent: CsvEntry = {
+		line1:    start?.line1 ?? 1,
+		col1:     start?.col1 ?? 1,
+		line2:    end?.line2 ?? 1,
+		col2:     end?.col2 ?? 1,
+		token:    RawRType.ExpressionList,
+		text:     '',
+		id:       RootId,
+		parent:   RootId,
+		terminal: false,
+		children: rootEntries,
+	}
+	// Return actual value.
+	return convertEntry(csvParent)
+}
+
+function convertEntry(csvEntry: CsvEntry): JsonEntry {
+	return {
+		...csvEntry,
+		// check and recursively iterate children
+		children: csvEntry.children?.sort(orderOf).map(convertEntry) ?? []
+	}
+}
+
+/**
+ * we sort children the same way xmlparsedata does (by line, by column, by inverse end line, by inverse end column, by terminal state, by combined "start" tiebreaker value)
+ * (https://github.com/r-lib/xmlparsedata/blob/main/R/package.R#L153C72-L153C78)
+ */
+function orderOf(c1: CsvEntry, c2: CsvEntry): number {
+	return c1.line1 - c2.line1 || c1.col1 - c2.col1 || c2.line2 - c1.line2 || c2.col2 - c1.col2 || Number(c1.terminal) - Number(c2.terminal) || sortTiebreak(c1) - sortTiebreak(c2)
```

Comment:
I am still thinking about if this would be cleaner with a spread in the parameters (i.e. `{line1, line2, col1, col2, terminal}: CsvEntry, ...`

Reply:
i don't think it would be at all, given there's two parameters with the same entries, so it would have to be, like, `line11`, `line21`, `line12`, `line22`, etc.